liferay-lima/liferay-portal
# LPS-190783 Bulk copying of folders, documents and file shortcuts

# What is this trying to solve?

https://issues.liferay.com/browse/LPS-190783

This PR includes an implementation of bulk copying and a significant refactor of the preexisting logic.

# How am I doing it?

The changes should be self explanatory. There were a lot of unnecessary differences and special cases; I've removed as much as I could, keeping all logic in a single view/action/display ctx.

# How can you verify that it works?

There are four cases that you need to check:
1. Copy a single document using the "Copy To" action in its dropdown.
2. Copy a single shortcut using the "Copy To" action in its dropdown.
3. Copy a single folder using the "Copy To" action in its dropdown.
4. Copy a mix of documents, folders and shortcuts using the management toolbar action.

Some things that I haven't fixed in this PR:
1. The way the action works when called from the dropdowns differ from the one in the management toolbar. In one case an intermediate view is displayed; in the other case it isnt. We should make all cases the same.
2. The frontend component needs to display multiple error messages.
3. The error messages need to include the element that failed.
4. The folder selector doesn't allow selecting the root folder.
